File: app_canibuythat/src/main/java/com/gb/canibuythat/ui/FileDialogActivity.java
package com.gb.canibuythat.ui;

import android.Manifest;
import android.app.AlertDialog;
import android.app.ListActivity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.gb.canibuythat.R;
import com.gb.canibuythat.util.DirReadException;
import com.gb.canibuythat.util.DirUtils;
import com.gb.canibuythat.util.PermissionVerifier;

import java.io.File;
import java.util.HashMap;
import java.util.List;

public class FileDialogActivity extends ListActivity {

    private static final int REQUEST_CODE_EXTERNAL_STORAGE_PERMISSION = 1;

    public static final int SELECTION_MODE_CREATE = 1;
    public static final int SELECTION_MODE_OPEN = 2;

    private static final String ROOT = "/";
    private static final String PARENT = "../";

    public static final String EXTRA_START_PATH = "START_PATH";
    public static final String EXTRA_FORMAT_FILTER = "FORMAT_FILTER";
    public static final String EXTRA_RESULT_PATH = "RESULT_PATH";
    public static final String EXTRA_SELECTION_MODE = "SELECTION_MODE";
    public static final String EXTRA_CAN_SELECT_DIR = "CAN_SELECT_DIR";

    private int selectionMode = SELECTION_MODE_CREATE;
    private String[] formatFilters = null;
    private boolean canSelectDir = false;

    private TextView pathView;
    private EditText fileNameView;
    private Button selectButton;
    private LinearLayout selectionButtonsContainer;
    private LinearLayout creationButtonsContainer;
    private InputMethodManager inputManager;

    private FileListAdapter adapter;
    private File clickedFile;
    private String currentPath = ROOT;
    private String parentPath;
    private HashMap<String, Integer> lastPositions = new HashMap<>();

    private PermissionVerifier permissionVerifier;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_file_picker);
        String[] permissions;
        if (selectionMode == SELECTION_MODE_CREATE) {
            permissions = new String[]{Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE};
        } else {
            permissions = new String[]{Manifest.permission.READ_EXTERNAL_STORAGE};
        }
        permissionVerifier = new PermissionVerifier(this, permissions);
        permissionVerifier.verifyPermissions(true, REQUEST_CODE_EXTERNAL_STORAGE_PERMISSION);

        setResult(RESULT_CANCELED, getIntent());

        selectionMode = getIntent().getIntExtra(EXTRA_SELECTION_MODE, SELECTION_MODE_CREATE);
        formatFilters = getIntent().getStringArrayExtra(EXTRA_FORMAT_FILTER);
        canSelectDir = getIntent().getBooleanExtra(EXTRA_CAN_SELECT_DIR, false);

        inputManager = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);

        pathView = (TextView) findViewById(R.id.path);
        fileNameView = (EditText) findViewById(R.id.file_name);
        selectButton = (Button) findViewById(R.id.select_btn);
        selectButton.setEnabled(false);
        selectButton.setOnClickListener(v -> {
            if (clickedFile != null) {
                getIntent().putExtra(EXTRA_RESULT_PATH, clickedFile.getPath());
                setResult(RESULT_OK, getIntent());
                finish();
            }
        });
        Button newButton = (Button) findViewById(R.id.new_btn);
        newButton.setOnClickListener(v -> {
            setCreateVisible(v);
            fileNameView.setText("");
            fileNameView.requestFocus();
        });
        if (selectionMode == SELECTION_MODE_OPEN) {
            newButton.setEnabled(false);
        }

        selectionButtonsContainer = (LinearLayout) findViewById(R.id.selection_buttons_container);
        creationButtonsContainer = (LinearLayout) findViewById(R.id.creation_buttons_container);
        creationButtonsContainer.setVisibility(View.GONE);
        Button cancelButton = (Button) findViewById(R.id.cancel_btn);
        cancelButton.setOnClickListener(this::setSelectVisible);
        Button createButton = (Button) findViewById(R.id.create_btn);
        createButton.setOnClickListener(v -> {
            if (fileNameView.getText().length() > 0) {
                getIntent().putExtra(EXTRA_RESULT_PATH, currentPath + "/" + fileNameView.getText());
                setResult(RESULT_OK, getIntent());
                finish();
            }
        });

        String startPath = getIntent().getStringExtra(EXTRA_START_PATH);
        startPath = startPath != null ? startPath : ROOT;
        if (canSelectDir) {
            clickedFile = new File(startPath);
            selectButton.setEnabled(true);
        }
        setData(startPath);
    }

    private void setData(String dirPath) {
        boolean useAutoSelection = dirPath.length() < currentPath.length();
        List<DirUtils.FileInfo> files = null;
        try {
            files = DirUtils.getDirInfo(dirPath, formatFilters);
        } catch (DirReadException e) {
            e.printStackTrace();
            Toast.makeText(FileDialogActivity.this, "Unable to read folder " + dirPath, Toast.LENGTH_SHORT).show();
            try {
                dirPath = ROOT;
                files = DirUtils.getDirInfo(dirPath, formatFilters);
            } catch (DirReadException e1) {
                e1.printStackTrace();
            }
        }
        parentPath = null;
        if (files != null) {
            if (!dirPath.equals(ROOT)) {
                files.add(0, new DirUtils.FileInfo(ROOT, true));
                files.add(1, new DirUtils.FileInfo(PARENT, true));
                parentPath = new File(dirPath).getParent();
            }

            currentPath = dirPath;
            adapter = new FileListAdapter(this, files);
            setListAdapter(adapter);

            Integer position = lastPositions.get(parentPath);
            if (position != null && useAutoSelection) {
                getListView().setSelection(position);
            }
            pathView.setText(getText(R.string.location) + ": " + currentPath);
        } else {
            setListAdapter(null);
            adapter = null;
            pathView.setText(null);
        }
    }

    private static class FileListAdapter extends ArrayAdapter<DirUtils.FileInfo> {

        FileListAdapter(Context context, List<DirUtils.FileInfo> items) {
            super(context, 0, items);
        }

        private class ViewHolder {
            ImageView mIcon;
            TextView mFilename;
        }

        @Override
        public
        @NonNull
        View getView(int position, View convertView, @NonNull ViewGroup parent) {
            View view;
            ViewHolder holder;

            if (convertView == null) {
                view = LayoutInflater.from(getContext()).inflate(R.layout.list_item_file_picker, parent, false);
                holder = new ViewHolder();
                holder.mIcon = (ImageView) view.findViewById(R.id.icon);
                holder.mFilename = (TextView) view.findViewById(R.id.file_name);
                view.setTag(holder);
            } else {
                view = convertView;
                holder = (ViewHolder) view.getTag();
            }
            DirUtils.FileInfo fileInfo = getItem(position);
            holder.mIcon.setImageResource(fileInfo.isFolder ? R.drawable.folder : R.drawable.file);
            holder.mFilename.setText(fileInfo.path);
            return view;
        }
    }

    @Override
    protected void onListItemClick(ListView l, View v, int position, long id) {
        String path = adapter.getItem(position).path;
        File file;
        if (path.equals(ROOT)) {
            file = new File(path);
        } else {
            file = new File(currentPath, path);
        }
        setSelectVisible(v);

        if (file.isDirectory()) {
            selectButton.setEnabled(false);
            if (file.canRead()) {
                lastPositions.put(currentPath, position);
                setData(file.getAbsolutePath());

                if (canSelectDir) {
                    clickedFile = file;
                    v.setSelected(true);
                    selectButton.setEnabled(true);
                }
            } else {
                new AlertDialog.Builder(this)
                        .setTitle("Folder [" + file.getName() + "] " + getText(R.string.cant_read_folder))
                        .setPositiveButton("OK", null).show();
            }
        } else {
            clickedFile = file;
            v.setSelected(true);
            selectButton.setEnabled(true);
        }
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if ((keyCode == KeyEvent.KEYCODE_BACK)) {
            selectButton.setEnabled(false);
            if (creationButtonsContainer.getVisibility() == View.VISIBLE) {
                creationButtonsContainer.setVisibility(View.GONE);
                selectionButtonsContainer.setVisibility(View.VISIBLE);
            } else {
                if (parentPath != null) {
                    setData(parentPath);
                } else {
                    return super.onKeyDown(keyCode, event);
                }
            }
            return true;
        } else {
            return super.onKeyDown(keyCode, event);
        }
    }

    private void setCreateVisible(View v) {
        creationButtonsContainer.setVisibility(View.VISIBLE);
        selectionButtonsContainer.setVisibility(View.GONE);
        inputManager.hideSoftInputFromWindow(v.getWindowToken(), 0);
        selectButton.setEnabled(false);
    }

    private void setSelectVisible(View v) {
        creationButtonsContainer.setVisibility(View.GONE);
        selectionButtonsContainer.setVisibility(View.VISIBLE);
        inputManager.hideSoftInputFromWindow(v.getWindowToken(), 0);
        selectButton.setEnabled(false);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (!permissionVerifier.onRequestPermissionsResult(requestCode, permissions, grantResults)) {
            Toast.makeText(this, "Missing permissions!", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app_canibuythat/src/main/java/com/gb/canibuythat/ui/FileDialogActivity.kt
package com.gb.canibuythat.ui

import android.Manifest
import android.app.Activity
import android.app.AlertDialog
import android.app.ListActivity
import android.content.Context
import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import com.gb.canibuythat.R
import com.gb.canibuythat.util.DirReadException
import com.gb.canibuythat.util.DirUtils
import com.gb.canibuythat.util.PermissionVerifier
import com.gb.canibuythat.util.hideKeyboard
import kotlinx.android.synthetic.main.activity_file_picker.*
import java.io.File
import java.util.*

class FileDialogActivity : ListActivity() {

    private var selectionMode = SELECTION_MODE_CREATE
    private var formatFilters: Array<String>? = null
    private var canSelectDir = false

    private var adapter: FileListAdapter? = null
    private var clickedFile: File? = null
    private var currentPath = ROOT
    private var parentPath: String? = null
    private val lastPositions = HashMap<String, Int>()

    private lateinit var permissionVerifier: PermissionVerifier

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_file_picker)
        val permissions: Array<String>
        if (selectionMode == SELECTION_MODE_CREATE) {
            permissions = arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE)
        } else {
            permissions = arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
        }
        permissionVerifier = PermissionVerifier(this, permissions)
        permissionVerifier.verifyPermissions(true, REQUEST_CODE_EXTERNAL_STORAGE_PERMISSION)

        setResult(Activity.RESULT_CANCELED, intent)

        selectionMode = intent.getIntExtra(EXTRA_SELECTION_MODE, SELECTION_MODE_CREATE)
        formatFilters = intent.getStringArrayExtra(EXTRA_FORMAT_FILTER)
        canSelectDir = intent.getBooleanExtra(EXTRA_CAN_SELECT_DIR, false)

        this.select_btn.isEnabled = false
        this.select_btn.setOnClickListener {
            clickedFile?.let {
                intent.putExtra(EXTRA_RESULT_PATH, it.path)
                setResult(Activity.RESULT_OK, intent)
                finish()
            }
        }
        val newButton = findViewById(R.id.new_btn) as Button
        newButton.setOnClickListener { v ->
            setCreateVisible(v)
            this.file_name_input.setText("")
            this.file_name_input.requestFocus()
        }
        if (selectionMode == SELECTION_MODE_OPEN) {
            newButton.isEnabled = false
        }

        this.selection_buttons_container.visibility = View.GONE
        this.cancel_btn.setOnClickListener(this::setSelectVisible)
        this.create_btn.setOnClickListener {
            if (this.file_name_input.text.isNotEmpty()) {
                intent.putExtra(EXTRA_RESULT_PATH, currentPath + "/" + this.file_name_input.text)
                setResult(Activity.RESULT_OK, intent)
                finish()
            }
        }

        var startPath: String? = intent.getStringExtra(EXTRA_START_PATH)
        startPath = startPath ?: ROOT
        if (canSelectDir) {
            clickedFile = File(startPath)
            this.select_btn.isEnabled = true
        }
        setData(startPath)
    }

    private fun setData(path: String) {
        var path = path
        val useAutoSelection = path.length < currentPath.length
        var files: MutableList<DirUtils.FileInfo>? = null
        try {
            files = DirUtils.getDirInfo(path, formatFilters)
        } catch (e: DirReadException) {
            // maybe it's a file
            val parent = File(path).parent

            try {
                files = DirUtils.getDirInfo(parent, formatFilters)
                setCreateVisible(null)
                this.file_name_input.setText(File(path).name)
                this.file_name_input.requestFocus()
                path = parent
            } catch (e2: DirReadException) {
                e.printStackTrace()
                Toast.makeText(this@FileDialogActivity, "Unable to read folder $path", Toast.LENGTH_SHORT).show()
                try {
                    path = ROOT
                    files = DirUtils.getDirInfo(path, formatFilters)
                } catch (e1: DirReadException) {
                    e1.printStackTrace()
                }
            }
        }

        parentPath = null
        files?.let {
            if (path != ROOT) {
                it.add(0, DirUtils.FileInfo(ROOT, true))
                it.add(1, DirUtils.FileInfo(PARENT, true))
                parentPath = File(path).parent
            }

            currentPath = path
            adapter = FileListAdapter(this, it)
            listAdapter = adapter

            val position = lastPositions[parentPath]
            if (position != null && useAutoSelection) {
                listView.setSelection(position)
            }
            this.path_lbl.text = "${getText(R.string.location)}: $currentPath"
        } ?: let {
            listAdapter = null
            adapter = null
            this.path_lbl.text = null
        }
    }

    private class FileListAdapter internal constructor(context: Context,
                                                       items: List<DirUtils.FileInfo>) : ArrayAdapter<DirUtils.FileInfo>(context, 0, items) {
        private inner class ViewHolder {
            lateinit var icon: ImageView
            lateinit var filename: TextView
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
            val view: View
            val holder: ViewHolder

            if (convertView == null) {
                view = LayoutInflater.from(context).inflate(R.layout.list_item_file_picker, parent, false)
                holder = ViewHolder()
                holder.icon = view.findViewById(R.id.icon) as ImageView
                holder.filename = view.findViewById(R.id.file_name_input) as TextView
                view.tag = holder
            } else {
                view = convertView
                holder = view.tag as ViewHolder
            }
            val fileInfo = getItem(position)
            holder.icon.setImageResource(if (fileInfo!!.isFolder) R.drawable.folder else R.drawable.file)
            holder.filename.text = fileInfo.path
            return view
        }
    }

    override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
        val path = adapter!!.getItem(position)!!.path
        val file: File
        file = if (path == ROOT) {
            File(path)
        } else {
            File(currentPath, path)
        }
        setSelectVisible(v)

        if (file.isDirectory) {
            this.select_btn.isEnabled = false
            if (file.canRead()) {
                lastPositions.put(currentPath, position)
                setData(file.absolutePath)

                if (canSelectDir) {
                    clickedFile = file
                    v.isSelected = true
                    this.select_btn.isEnabled = true
                }
            } else {
                AlertDialog.Builder(this)
                        .setTitle("Folder [" + file.name + "] " + getText(R.string.cant_read_folder))
                        .setPositiveButton("OK", null).show()
            }
        } else {
            clickedFile = file
            v.isSelected = true
            this.select_btn.isEnabled = true
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            this.select_btn.isEnabled = false
            if (this.creation_buttons_container.visibility == View.VISIBLE) {
                this.creation_buttons_container.visibility = View.GONE
                this.selection_buttons_container.visibility = View.VISIBLE
            } else {
                parentPath?.let {
                    setData(it)
                } ?: let {
                    return super.onKeyDown(keyCode, event)
                }
            }
            return true
        } else {
            return super.onKeyDown(keyCode, event)
        }
    }

    private fun setCreateVisible(v: View?) {
        this.creation_buttons_container.visibility = View.VISIBLE
        this.selection_buttons_container.visibility = View.GONE
        v?.apply {
            hideKeyboard()
        }
        this.select_btn.isEnabled = false
    }

    private fun setSelectVisible(v: View?) {
        this.creation_buttons_container.visibility = View.GONE
        this.selection_buttons_container.visibility = View.VISIBLE
        v?.apply {
            hideKeyboard()
        }
        this.select_btn.isEnabled = false
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (!permissionVerifier.onRequestPermissionsResult(requestCode, permissions, grantResults)) {
            Toast.makeText(this, "Missing permissions!", Toast.LENGTH_SHORT).show()
        }
    }

    companion object {
        private val REQUEST_CODE_EXTERNAL_STORAGE_PERMISSION = 1

        val SELECTION_MODE_CREATE = 1
        val SELECTION_MODE_OPEN = 2

        private val ROOT = "/"
        private val PARENT = "../"

        val EXTRA_START_PATH = "START_PATH"
        val EXTRA_FORMAT_FILTER = "FORMAT_FILTER"
        val EXTRA_RESULT_PATH = "RESULT_PATH"
        val EXTRA_SELECTION_MODE = "SELECTION_MODE"
        val EXTRA_CAN_SELECT_DIR = "CAN_SELECT_DIR"
    }
}